# Advent of Code 2020

### Kwame V. Taylor

---

#### Day 2: Password Philosophy

Your flight departs in a few days from the coastal airport; the easiest way down to the coast from here is via toboggan.

The shopkeeper at the North Pole Toboggan Rental Shop is having a bad day. "Something's wrong with our computers; we can't log in!" You ask if you can take a look.

Their password database seems to be a little corrupted: some of the passwords wouldn't have been allowed by the Official Toboggan Corporate Policy that was in effect when they were chosen.

To try to debug the problem, they have created a list (your puzzle input) of **passwords** (according to the corrupted database) and **the corporate policy when that password was set**.

For example, suppose you have the following list:

```
1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc
```

Each line gives the password policy and then the password. The password policy indicates the lowest and highest number of times a given letter must appear for the password to be valid. For example, ```1-3 a``` means that the password must contain ```a``` at least ```1``` time and at most ```3``` times.

In the above example, ```2``` passwords are valid. The middle password, ```cdefg```, is not; it contains no instances of ```b```, but needs at least ```1```. The first and third passwords are valid: they contain one ```a``` or nine ```c```, both within the limits of their respective policies.

**How many passwords are valid** according to their policies?

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('input.txt', sep=': ', engine='python', header=None)

In [3]:
df = df.rename(columns={0: 'policy', 1: 'password'})

In [4]:
df.head()

policy            password
0   4-8 n  dnjjrtclnzdnghnbnn
1   5-6 r             rrrrcqr
2   2-4 p         pqhpppwpfmr
3  1-13 p      spppppppppppvp
4   5-6 x            xfwxxxxx

In [5]:
df.shape

(1000, 2)

I need to translate each of the policy requirements into regex to search for within each password.

In [6]:
# Python3 code to demonstrate  
# occurrence frequency using  
# collections.Counter() 
from collections import Counter 
  
# initializing string  
test_str = "dnjjrtclnzdnghnbnn"
  
# using collections.Counter() to get count  
# counting e  
count = Counter(test_str) 
  
# printing result  
print ("Count of n in dnjjrtclnzdnghnbnn is : "
                       +  str(count['n'])) 

Count of n in dnjjrtclnzdnghnbnn is : 6


In [7]:
'dnjjrtclnzdnghnbnn'.count('n')

6

In [8]:
df[['policy_range','policy_letter']] = df.policy.apply(lambda x: pd.Series(str(x).split(" ")))

In [9]:
df = df.drop(columns=['policy'])
df.head()

password policy_range policy_letter
0  dnjjrtclnzdnghnbnn          4-8             n
1             rrrrcqr          5-6             r
2         pqhpppwpfmr          2-4             p
3      spppppppppppvp         1-13             p
4            xfwxxxxx          5-6             x

In [10]:
#df.policy_range = df.policy_range.str.replace('-', ' ')
#df.policy_range = df.policy_range.str.split(" ").apply(tuple)
#df.policy_range = tuple(df.policy_range)

In [11]:
#df.head()

In [12]:
#df.policy_range.dtype

In [13]:
#df.policy_letter = ("'" + (df.policy_letter).astype(str) + "'")

In [14]:
df.head()

password policy_range policy_letter
0  dnjjrtclnzdnghnbnn          4-8             n
1             rrrrcqr          5-6             r
2         pqhpppwpfmr          2-4             p
3      spppppppppppvp         1-13             p
4            xfwxxxxx          5-6             x

In [15]:
#df['is_valid'] = df['password'].count(df['policy_letter'])

In [16]:
df['counter'] = df.apply(lambda x: x.password.count(x.policy_letter), axis=1)

In [17]:
df.head()

password policy_range policy_letter  counter
0  dnjjrtclnzdnghnbnn          4-8             n        6
1             rrrrcqr          5-6             r        5
2         pqhpppwpfmr          2-4             p        5
3      spppppppppppvp         1-13             p       12
4            xfwxxxxx          5-6             x        6

In [18]:
df[['policy_range_start','policy_range_end']] = df.policy_range.apply(lambda x: pd.Series(str(x).split("-")))

In [19]:
df = df.drop(columns=['policy_range'])
df.head()

password policy_letter  counter policy_range_start  \
0  dnjjrtclnzdnghnbnn             n        6                  4   
1             rrrrcqr             r        5                  5   
2         pqhpppwpfmr             p        5                  2   
3      spppppppppppvp             p       12                  1   
4            xfwxxxxx             x        6                  5   

  policy_range_end  
0                8  
1                6  
2                4  
3               13  
4                6

In [20]:
df.dtypes

password              object
policy_letter         object
counter                int64
policy_range_start    object
policy_range_end      object
dtype: object

In [21]:
df.policy_range_start = pd.to_numeric(df.policy_range_start, errors='coerce').fillna(0).astype(np.int64)
df.policy_range_end = pd.to_numeric(df.policy_range_end, errors='coerce').fillna(0).astype(np.int64)

In [22]:
df.dtypes

password              object
policy_letter         object
counter                int64
policy_range_start     int64
policy_range_end       int64
dtype: object

In [25]:
df.query('counter >= policy_range_start & counter <= policy_range_end')

password policy_letter  counter  policy_range_start  \
0      dnjjrtclnzdnghnbnn             n        6                   4   
1                 rrrrcqr             r        5                   5   
3          spppppppppppvp             p       12                   1   
4                xfwxxxxx             x        6                   5   
5    nnnnsnnnnnnnnnnnnnnn             n       19                  19   
..                    ...           ...      ...                 ...   
991           qqrqqqqqlkq             q        8                   8   
993       ssssssssscsfsss             s       13                   8   
994          pwwtwwlwwhww             w        8                   6   
996             jvgxmphvr             v        2                   2   
997                rrzrcd             r        3                   2   

     policy_range_end  
0                   8  
1                   6  
3                  13  
4                   6  
5                  20  
..                ...  
991                11  
993                13  
994                 8  
996                 4  
997                 6  

[622 rows x 5 columns]

622 passwords are valid!

> ```That's the right answer! You are one gold star closer to saving your vacation.```

#### Part Two

While it appears you validated the passwords correctly, they don't seem to be what the Official Toboggan Corporate Authentication System is expecting.

The shopkeeper suddenly realizes that he just accidentally explained the password policy rules from his old job at the sled rental place down the street! The Official Toboggan Corporate Policy actually works a little differently.

Each policy actually describes **two positions in the password**, where ```1``` means the first character, ```2``` means the second character, and so on. (Be careful; Toboggan Corporate Policies have no concept of "index zero"!) **Exactly one of these positions** must contain the given letter. Other occurrences of the letter are irrelevant for the purposes of policy enforcement.

Given the same example list from above:

```1-3 a: abcde``` is **valid**: position ```1``` contains ```a``` and position ```3``` does not.
```1-3 b: cdefg``` is **invalid**: neither position ```1``` nor position ```3``` contains ```b```.
```2-9 c: ccccccccc``` is **invalid**: both position ```2``` and position ```9``` contain ```c```.

**How many passwords are valid** according to the new interpretation of the policies?